RFC: Split Platform and Connector Builds

cgardens opened the pull request:
Closes https://github.com/airbytehq/airbyte/issues/3828

## What
When developing on the platform, builds are frequently fouled up by build issues related to parts of the connector base and formatting for python sub projects. The opposite is true as well. The most common cases are that one of the many dependencies for running format on python connectors fails transiently or in general that formatting gets fouled up in some way when connectors are contributed.

Our goal is to allow developers to iterate autonomously and efficiently in different parts of the code base.

## How
Doubles down on the approach we already use to split builds. In `settings.gradle` it removes projects that are not relevant based on an environment variable. There is a tradeoff here. This approach is _not_ considered a best practice. 

The reason I wanted to explore it is that it leads to a deeper separation that other approaches. In other approaches we can tag projects in gradle as part of platform or whatever, but those tags would all be processed during the `configuration` phase in gradle. This means that gradle problems in connectors still block progress in other parts of the code base. One can argue that this isn't the main case we are trying to address and so going with a mediocre practice isn't worth it.

Definitely happy to go ahead with the other approach I mentioned here if the consensus is that we should stick with the better practice. Either way I wanted to get the ball rolling and give  a sense of what the split build would look like in the CI.

## Recommended reading order
1. `settings.gradle`
2. `.github/workflows/gradle.yml`

heads up @jrhizor @sherifnada

cgardens commented on airbyte-config/init/src/main/resources/config/STANDARD_SOURCE_DEFINITION/41375467-61ae-4204-8e38-e2b8b7365f23.json:
```diff
@@ -1,6 +1,6 @@
 {
   "sourceDefinitionId": "41375467-61ae-4204-8e38-e2b8b7365f23",
-  "name": "Slack",
+  "name": "Slack (Singer)",
```
ignore. noise.

cgardens commented on airbyte-config/init/src/main/resources/config/STANDARD_SOURCE_DEFINITION/c2281cee-86f9-4a86-bb48-d23286b4c7bd.json:
```diff
@@ -1,6 +1,6 @@
 {
   "sourceDefinitionId": "c2281cee-86f9-4a86-bb48-d23286b4c7bd",
-  "name": "Slack CDK",
```
ignore. noise.

cgardens commented on airbyte-config/init/src/main/resources/seed/source_definitions.yaml:
```diff
@@ -161,7 +161,7 @@
   documentationUrl: https://hub.docker.com/r/airbyte/source-braintree-singer
   icon: braintree.svg
 - sourceDefinitionId: 41375467-61ae-4204-8e38-e2b8b7365f23
-  name: Slack
```
ignore. noise.

jrhizor commented:
Definitely don't love it, but it looks like a good short term workaround.

We need to completely cut off connector builds from the main build at some point. We shouldn't do configuration for hundreds of connector builds.

sherifnada commented:
Seems like a decent stop gap but at this point we are stacking those unfortunately. Agree w Jared. We should fix the build situation more sustainably in August.

sherifnada commented:
Also can you add some docs so the different subcommands are documented somewhere?

cgardens commented:
Thanks. I agree with everything that as been said.

Yeah. I think the next step here is to figure out how to more formally split connectors and platform. Like I don't think there's some gradle feature we just haven't discovered yet. Next step is a code organization one. Dunno if that's splitting the monorepo or just having 2 gradle builds in one project or what.

davinchia reviewed (approved):
Not the worst approach - looks good!

Can we have a quick summary in `gradle-cheatsheet.md` for future devs?

I know there are various gradlew invocations in the tools/bin directory - should we also updated those?

Non-blocking from me.

cgardens commented:
@davinchia @sherifnada I added documentation in `gradle-cheatsheet.md`. 

@davinchia updated all usages of `./gradlew` in the project to use a sub build where it makes sense to do so.

My plan is to merge this tomorrow morning so that I can be around if it causes any developer problems (which I'm sure it will have some rough edges at the very start).

cgardens commented:
I am still fighting the connector template generator in the CI. I can't get it to run locally on master either. The irony here is the whole point of this split is to avoid this sort of problem. I am going to merge this as is now to avoid needing to handle continuous drift and so that the main change can be socialized during sane hours, but will keep plugging away on trying to fix the generator problem.

cgardens commented:
For posterity here is what the build times look like before and after. Interestingly the split builds when summed up are still faster than the combined one. My guess is because the split builds are better at not redoing unnecessary work. For example, we should not longer ever be formatting every single connector.

Original combined build times.
total: 25 min

![Screen Shot 2021-07-15 at 10 39 00 AM](https://user-images.githubusercontent.com/9092207/125833472-a98f72be-f529-4b31-bb51-19fb1a887872.png)

Split Connectors Build (without scaffold which will add a 1-2 min)
total: 4-5 min
![Screen Shot 2021-07-15 at 10 39 44 AM](https://user-images.githubusercontent.com/9092207/125833508-c8c1872f-dc10-4229-adf2-51100f9d4dd8.png)

total: 8-9 min
Split Platform Build
![Screen Shot 2021-07-15 at 10 45 18 AM](https://user-images.githubusercontent.com/9092207/125833561-fcd84995-3502-4d05-9140-42d403bb9895.png)
